doc/engineering/coding_standards.md

```diff
@@ -0,0 +1,236 @@
+  * critical: The system (or a part of it) crashed and somebody should be informed and take action about it
+* Simple unit tests
+  *Single responsibility applies to tests as well. When writing unit tests, keep them short and don't put more than 1 or 2 asserts per test
+
+Some 2600Hz specifics:
+
+* Tickie all atoms
+  * Use `'foo'` not `foo`
+* Write accurate specs for all functions
+  * Use the most restrictive type(s) you can.
+  ```erlang
+  -spec foo(any()) -> any(). % is of no use to anyone!
+  ```
+  * Run [dialyzer](./dialyzer.md) on any modules you make changes to, plus any modules your new code makes calls to.
+* When creating lists or binaries, drop the comma and next element to a new line
+  * This includes `-export([...])` directives.
+  * No:
+  ```erlang
+  [this, list, is, wrong]
+  ```
+  * Yes:
+  ```erlang
+  [this
+  ,list
+  ,is
+  ,right
+  ]
+  ```
+* Exports should be grouped by functionality.
+  * Do not export all functions nor add all exports to a single export directive.
+* Code should be self documenting.
+  * It is not a forum for you to prove how amazing you are at the language.
+  * We always prefer readability over complexity. Which means simple language constructs and explicit variable/function names.
+    * `AccountId` instead of `Id` or `AcctId`
+* The function name should clearly describe what it will do.
+  * Common patterns include:
+    * The use of plural to singular.  For example, `process_accounts` calls to `process_account` for each account.
+    * The use of `maybe_X` executing `X` if a condition is met.  For example, `maybe_process_accounts` calls to `process_accounts` if there are accounts present in the system.
+* Never introduce a single letter variable.
+  * It is marginally acceptable to use a single letter in “inner” scope variables but they should match the first letter of a fully named variable they are shadowing. For example, in the predicate function it is OK to use `A` to represent `Account` if the `Account` variable is used outside the predicate.
+* Do not chain functions
+  * Do not call the next function of a sequence of functions from the previous function.
+  * If you need to chain functions together to make a sequence fold over a list of functions.
+  * No:
+  ```erlang
+  do_x() ->
+      X = lookup(),
+      do_y(X).
+  do_y(X) ->
+      Y = transform_x(X),
+      do_z(Y).
+  do_z(Y) ->
+      apply_z(Y).
+  ```
+  * Yes:
+  ```erlang
+  do_x() ->
+      lists:foldl(fun(F, Acc) -> F(Acc) end, do_x(), [fun do_y/1, do_z/1]).
+  ```
+
+Some additional guidance can be found:
+* http://www.erlang.se/doc/programming_rules.shtml
+* https://www.youtube.com/watch?v=CQyt9Vlkbis
+
+Some in-progress work is using [Elvis](https://github.com/inaka/elvis) to enforce style requirements so you can get feedback automatically for where your code has lost its way. This will eventually be added to Travis as well.
```

Comment:
Please see for yourself https://github.com/2600hz/kazoo/blob/master/.travis.yml#L50